src/ophyd_async/epics/adcore/_hdf_writer.py

```diff
@@ -75,6 +76,19 @@ async def open(
         # Determine number of frames that will be saved per HDF chunk
         frames_per_chunk = await self.fileio.num_frames_chunks.get_value()
 
+        if None in detector_shape:
+            # Questions:
+            # - Can AreaDetector support this?
+            # - How to deal with chunking?
+            # Don't support for now - leave option open to support it later
+            raise ValueError(
+                "Datasets with partially unknown dimensionality "
+                "are not currently supported by ADHDFWriter."
+            )
+
+        # Pyright doesn't understand the above check as a type-narrowing expression
+        detector_shape = cast(tuple[int, ...], detector_shape)
```

Comment:
```suggestion
      def is_fully_described(val: Tuple[int | None, ...]) -> TypeGuard[tuple[int, ...]]:
            # Questions:
            # - Can AreaDetector support this?
            # - How to deal with chunking?
            # Don't support for now - leave option open to support it later
            return None not in val

        if not is_fully_described():
            raise ValueError(
                "Datasets with partially unknown dimensionality "
                "are not currently supported by ADHDFWriter."
            )
```


I haven't used it, but I think this is how to tell Pyright that you're narrowing the permitted types
https://docs.python.org/3/library/typing.html#typing.TypeGuard